File: backend/schemas/profit.py
from pydantic import BaseModel, ConfigDict, field_validator

class ProfitLineResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    stock_id: str
    name: str
    quantity: int
    retail_value: int
    cost_value: int
    profit: int | None
    priced: bool
    costed: bool

    @field_validator("stock_id", mode="before")
    @classmethod
    def coerce_id(cls, value: object) -> str:
        return str(value)

class ProfitSummaryResponse(BaseModel):
    total_retail_value: int
    total_cost_value: int
    unrealized_profit: int
    priced_assets: int
    costed_assets: int
    complete_assets: int
    uncosted_assets: int
    by_origin: dict[str, int]
    by_genre: dict[str, int]
    lines: list[ProfitLineResponse]

File: frontend/src/types/Profit.ts
export type ProfitLine = {
  stock_id: string;
  name: string;
  quantity: number;
  retail_value: number;
  cost_value: number;
  profit: number | null;
  priced: boolean;
  costed: boolean;
};

export type ProfitSummary = {
  total_retail_value: number;
  total_cost_value: number;
  unrealized_profit: number;
  priced_assets: number;
  costed_assets: number;
  complete_assets: number;
  uncosted_assets: number;
  by_origin: Record<string, number>;
  by_genre: Record<string, number>;
  lines: ProfitLine[];
};